sequencer/src/block.rs

```diff
@@ -17,28 +19,131 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
-#[derive(Default)]
+// New Type for `U256` in order to implement `CanonicalSerialize` and
+// `CanonicalDeserialize`
+#[derive(Default, Hash, Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct FeeElement(U256);
+// New Type for `Address` in order to implement `CanonicalSerialize` and
+// `CanonicalDeserialize`
+#[derive(
+    Default, Hash, Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord,
+)]
+pub struct FeeIndex(Address);
+
+impl Add for FeeElement {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self(self.0.add(other.0))
+    }
+}
+
+impl Valid for FeeElement {
+    fn check(&self) -> Result<(), SerializationError> {
+        Ok(())
+    }
+}
+
+impl Valid for FeeIndex {
+    fn check(&self) -> Result<(), SerializationError> {
+        Ok(())
+    }
+}
+
+impl CanonicalSerialize for FeeElement {
+    fn serialize_with_mode<W: std::io::prelude::Write>(
+        &self,
+        mut writer: W,
+        _compress: Compress,
+    ) -> Result<(), SerializationError> {
+        let mut bytes = [0u8; core::mem::size_of::<U256>()];
+        self.0.to_little_endian(&mut bytes);
+        Ok(writer.write_all(&bytes)?)
+    }
+
+    fn serialized_size(&self, _compress: Compress) -> usize {
+        core::mem::size_of::<U256>()
+    }
+}
+impl CanonicalDeserialize for FeeElement {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        _compress: Compress,
+        _validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let mut bytes = [0u8; core::mem::size_of::<U256>()];
+        // let mut ret = [0u8; 32];
+        reader.read_exact(&mut bytes)?;
+        let value = U256::from_little_endian(&bytes);
+        Ok(Self(value))
+    }
+}
+impl CanonicalSerialize for FeeIndex {
+    fn serialize_with_mode<W: std::io::prelude::Write>(
+        &self,
+        mut writer: W,
+        _chompress: Compress,
+    ) -> Result<(), SerializationError> {
+        Ok(writer.write_all(self.0.as_bytes())?)
+    }
+
+    fn serialized_size(&self, _compress: Compress) -> usize {
+        core::mem::size_of::<Address>()
+    }
+}
+impl CanonicalDeserialize for FeeIndex {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        _compress: Compress,
+        _validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let mut bytes = [0u8; core::mem::size_of::<Address>()];
+        reader.read_exact(&mut bytes)?;
+        let value = Address::from_slice(&bytes);
+        Ok(Self(value))
+    }
+}
+impl std::convert::From<u64> for FeeIndex {
+    fn from(item: u64) -> Self {
+        FeeIndex(Address::from_low_u64_le(item))
+    }
+}
+
+impl<U2: Unsigned> ToTraversalPath<U2> for FeeIndex {
+    fn to_traversal_path(&self, height: usize) -> Vec<usize> {
+        let mut pos = Address::to_low_u64_le(&self.0);
+        let mut ret = vec![];
+        for _i in 0..height {
+            let x = (pos % (U2::to_u64())) as usize;
```

Comment:
Since the arity is 2 here. We may replace this repeated mod and division by a simple `to_le_bits` function. However, I could only find `to_le_bytes`.

This is already good enough though.